spec/event_store_client/adapters/grpc/commands/command_spec.rb

```diff
@@ -0,0 +1,181 @@
+
+  describe '.use_request' do
+    subject { described_class.use_request(request_class) }
+
+    let(:request_class) { Class.new }
+
+    it 'registers request class' do
+      expect { subject }.to change {
+        EventStoreClient::GRPC::CommandRegistrar.request(described_class)
+      }.to(request_class)
+    end
+  end
+
+  describe '.use_service' do
+    subject { described_class.use_service(service_class) }
+
+    let(:service_class) { Class.new }
+
+    it 'registers request class' do
+      expect { subject }.to change {
+        EventStoreClient::GRPC::CommandRegistrar.service(described_class)
+      }.to(service_class)
+    end
+  end
+
+  describe '#call' do
+    subject { instance.call }
+
+    it { expect { subject }.to raise_error(NotImplementedError) }
+  end
+
+  describe '#metadata' do
+    subject { instance.metadata }
+
+    let(:conn_options) { { username: 'some-user', password: '1234'} }
+
+    describe 'when connection is secure' do
+      before do
+        EventStoreClient.config.eventstore_url.tls = true
+      end
+
+      it 'includes credentials into metadata' do
+        credentials =
+          Base64.
+            encode64("#{conn_options[:username]}:#{conn_options[:password]}").
+            delete("\n")
+        is_expected.to eq('authorization' => "Basic #{credentials}")
+      end
+    end
+
+    describe 'when connection is insecure' do
+      it { is_expected.to eq({}) }
+    end
+  end
+
+  describe '#request' do
+    subject { instance.request }
+
+    let(:request_class) { Class.new }
+
+    before do
+      described_class.use_request(request_class)
+    end
+
+    it 'returns request class' do
+      is_expected.to eq(request_class)
+    end
+  end
+
+  describe '#service' do
+    subject { instance.service }
+
+    let(:service_class) { EventStore::Client::Gossip::Gossip::Stub }
+
+    before do
+      described_class.use_service(service_class)
+    end
+
+    it 'returns instance of service class' do
+      is_expected.to be_an_instance_of(service_class)
+    end
+  end
+
+  describe '#connection_options' do
+    subject { instance.connection_options }
+
+    it { is_expected.to eq(instance.send(:connection).options_hash) }
+  end
+
+  describe '#retry_request' do
+    subject { instance.send(:retry_request, &blk) }
+
+    let(:blk) { proc { result } }
+    let(:result) { 'some-result' }
+
+    context 'when no errors raises' do
+      it 'returns result' do
+        is_expected.to eq(result)
+      end
+    end
+
+    context 'when raises error' do
+      let(:result) { request.call }
+      let(:request) { double('Request') }
+      let(:error_class) { GRPC::Unavailable }
+
+      before do
+        allow(request).to receive(:call).and_raise(error_class)
+      end
+
+      context 'when GRPC::Unavailable error raises' do
+        it 'retries request up to config.eventstore_url.grpc_retry_attempts times' do
+          begin
+            subject
+          rescue error_class
```

Comment:
To prevent raising it if the implementation changes. We are dealing with conditional `rescue`/`raise`, so I would like a developer to see the test error of what I am testing there, and not an exception